src/mito2/src/memtable/bulk/part.rs

```diff
@@ -0,0 +1,666 @@
+                check_bounds: false,
+            }),
+        )
+        .context(ComputeArrowSnafu)?,
+    );
+
+    arrays.push(
+        arrow::compute::take(
+            &op_type,
+            &indices,
+            Some(TakeOptions {
+                check_bounds: false,
+            }),
+        )
+        .context(ComputeArrowSnafu)?,
+    );
+
+    let batch = RecordBatch::try_new(schema, arrays).context(NewRecordBatchSnafu)?;
+    Ok((batch, min_timestamp, max_timestamp))
+}
+
+/// Converts timestamp array to an iter of i64 values.
+fn timestamp_array_to_iter<'a>(
+    metadata: &RegionMetadataRef,
+    timestamp: &'a ArrayRef,
+) -> impl Iterator<Item = &'a i64> {
+    // safety: timestamp column must be valid, and values must not be None.
+    match metadata
+        .time_index_column()
+        .column_schema
+        .data_type
+        .as_timestamp()
+        .unwrap()
+    {
+        // safety: timestamp column must be valid.
+        TimestampType::Second(_) => timestamp
+            .as_any()
+            .downcast_ref::<TimestampSecondArray>()
+            .unwrap()
+            .values()
+            .iter(),
+        TimestampType::Millisecond(_) => timestamp
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
+            .unwrap()
+            .values()
+            .iter(),
+        TimestampType::Microsecond(_) => timestamp
+            .as_any()
+            .downcast_ref::<TimestampMicrosecondArray>()
+            .unwrap()
+            .values()
+            .iter(),
+        TimestampType::Nanosecond(_) => timestamp
+            .as_any()
+            .downcast_ref::<TimestampNanosecondArray>()
+            .unwrap()
+            .values()
+            .iter(),
+    }
+}
+
+/// Converts a **sorted** [BinaryArray] to [DictionaryArray].
+fn binary_array_to_dictionary(input: &BinaryArray) -> Result<DictionaryArray<UInt16Type>> {
+    if input.is_empty() {
+        return Ok(DictionaryArray::new(
+            <UInt16Array as From<Vec<u16>>>::from(vec![]),
+            Arc::new(BinaryArray::from_vec(vec![])) as ArrayRef,
+        ));
+    }
+    let mut keys = UInt16BufferBuilder::new(16);
+    let mut values = BinaryBuilder::new();
+    let mut prev: usize = 0;
+    keys.append(prev as u16);
+    values.append_value(input.value(prev));
+
+    for current_bytes in input.iter().skip(1) {
+        // safety: encoded pk must present.
+        let current_bytes = current_bytes.unwrap();
+        let prev_bytes = input.value(prev);
+        if current_bytes != prev_bytes {
+            values.append_value(current_bytes);
+            prev += 1;
+        }
+        keys.append(prev as u16);
+    }
+
+    Ok(DictionaryArray::new(
+        UInt16Array::from(
+            ArrayData::builder(ArrowDataType::UInt16)
+                .len(keys.len())
+                .add_buffer(keys.finish())
+                .build()
+                .context(ComputeArrowSnafu)?,
+        ),
```

Comment:
Consider using `From<Vec<u16>>`. We can define `keys` as `Vec<u16>`. `Buffer` could be constructed from `Vec` without copying.
https://docs.rs/arrow/latest/arrow/array/struct.PrimitiveArray.html#example-from-a-vec
https://docs.rs/arrow/latest/arrow/buffer/struct.Buffer.html#method.from_vec